Get more content and transform helpers

JulieWinchester opened the pull request:
This PR adds or changes the following functionality: 
* Canvas and Scene expose non-content annotations (listed in `annotations` property instead of `items`) through method `getNonContentAnnotations()`
* Camera `getLookAt` method can now return a SpecificResource in addition to the PointSelector or Annotation URIs previously supported. This allows for a camera to look at points referenced outside the current Scene scope.
* SpecificResource exposes scope information (provided in 'scope' property) through method `getScope()`. The `scope` property is from WADM but has not been used significantly in IIIF previously, but will be mentioned in spec documents and used moving forward for providing annotation scope content state information (i.e., here is the state of the viewer that provides the best context with which to view this annotation in terms of camera placement, etc.). 
* Annotation scope content state information is exposed from the Annotation through the `getScopeContent()` method. This method assumes the Annotation has a target which is a Specific Resource, and that the Specific Resource target has `scope` specified, and that the scope is associated with a list of subsequent annotations. The method returns these scope content state annotations. This occurs in a situation where an Annotation's target has a scope property which is a content state with a list of annotation items, where a client is to understand that these scope annotation items are important for understanding the original Annotation. This allows downstream packages to easily access through scope content state annotation items. 
* Adds transform helper methods to Geometry3D. In particular, there are two methods to help downstream packages get "end state" combinations of transforms, `combineTransformsToMatrix()` and `combineTransformsToTRS()`, which use different methods to combine transforms and **importantly, produce different results**. The method `combineTransformsToMatrix()` builds a 4x4 transformation matrix out of a sequence of Transforms in order and returns the transformation matrix. This method **does not** produce combined transforms in the way the Presentation spec dictates they should be combined, where each transformation is carried out on the local coordinate set or space and the output of that transformation is the input of the next set. The method `combineTransformsToTRS()` **does** produce combined transforms how the spec expects them to be combined, and returns a set of transforms with objects for translation, rotation, and scale. Probably the first method should be removed, but since this is how we discussed handling it in the IIIF commons meeting, it's still there for now. 
* Adds convenience methods to AnnotationBody to get the combined end state transform using the Geometry3D methods above, through methods `getTransformMatrix()` and `getTransformSet` (which uses `combineTransformsToTRS()`).

JulieWinchester committed:
Add 3D transform helper methods

JulieWinchester committed:
Merge branch 'master' into transforms_wip

JulieWinchester committed:
Add ability to retrieve non-content annotations from Canvas/Scene and scope content state annotation items from Annotation

JulieWinchester committed:
Support lookAt being SpecificResource

JulieWinchester committed:
Fix bug in getting scope from specific resource

JulieWinchester committed:
Add better method of getting 3D transforms that corresponds more to IIIF expectations

JulieWinchester committed:
Merge branch 'master' into get_more_content_and_transform_helpers

vincentmarchetti committed:
Removed expect commands tests that were causing false failures
camera tests were testing for an explicit value of fieldOfView property which
are not defined in these individual manifests

vincentmarchetti commented:
At commit https://github.com/IIIF-Commons/manifesto-3d/pull/29/commits/8d6ffe3ed147d75ae4e125ca73e203fafe9f25f5 running the unit tests
`npm run test`
gave 4 failures in the 3d-related tests. 3 of these failures were because the fieldOfView property of a camera was being tested against an exact value (45 degrees) while in the manifests being queried the fieldOfView property was not defined. In the Pres 4 preview draft the fieldOfView property is optional (at the SHOULD level, not MUST)  so I judged these are false-positive failures. At commit https://github.com/IIIF-Commons/manifesto-3d/pull/29/commits/9b783333ab5aa1f5402c542f018235e6a38ec3d2 I have removed these fieldOfView tests.

The remaining failure will be discussed in a following comment.

vincentmarchetti commented:
At commit https://github.com/IIIF-Commons/manifesto-3d/commit/9b783333ab5aa1f5402c542f018235e6a38ec3d2 one 3d-related unit test is failing:

```
1) iiif-3d tests
       10_content_state
         astronaut_comment_scope
           astronaut_comment_scope
             with camera annotation:

      AssertionError: expected 0 to equal 1
      + expected - actual

      -0
      +1
      
      at Context.<anonymous> (test/tests_3d/10_content_state/astronaut_comment_scope.js:90:43)
      at process.processImmediate (node:internal/timers:478:21)
```

It is failing because the getScopeContent method applied to a commenting annotation is not returning any of the painting annotations in the target>scope>target (scene) > items. It looks like this is because the getScopeContent method in Annotation.ts does not take into account that the painting annotation resource may be (or perhaps must be) contained in an AnnotationPage collection.

vincentmarchetti commented:
The source file `src/geometry3d.ts` declare a Typescript type TransformSet. From the way it is constructed from a list of IIIF Transform resources it looks like the intended action of a TransformSet on a set of 3 x,y,z coordinates in a local coordinate system, to yield x,y,z coordinates in the global coordinates, is, in following order

1. Rotated and scale the three coordinates, by Euler calculation and vector multiplication
2. Translate the three coordinates by vector addition

Or, in language of threejs-math, where tset is a TransformSet object, and v is a threejs Vector3 :
```
+function applyTransformSet( tset, v )  {
+        return v.clone().applyEuler(tset.rotation).multiply(tset.scale).add(tset.translation);        
+};
```

The multiplication of threejs-math Matrix4 objects in the `combineTransformsToMatrix `is not being done in the correct order, we need to make this change:

```
--- a/src/Geometry3d.ts
+++ b/src/Geometry3d.ts
@@ -126,7 +126,7 @@ export function combineTransformsToMatrix(transforms: Transform[]): Matrix4 {
             const scale: any = (transform as ScaleTransform).getScale();
             tmat.makeScale(scale.x, scale.y, scale.z);
         }
-        matrix.multiply(tmat);
+        matrix.premultiply(tmat);
     }

```
The code in the `combineTransformsToTRS `function that combines two  Euler rotations needs to be recoded convert both to quaternions and using one of the Quaternion *multiply functions.

If these are done, I think we can write a suite of test cases that will show the same calculation of coordinates local -> global can be done equivalently with the TransformSet methods or the Matrix4 methods. Clients may find it useful to have these two options.

JulieWinchester committed:
Fix Annotation getScopeContent method to handle Annotation Pages

JulieWinchester commented:
Thanks for the review @vincentmarchetti! I've updated Annotation's  `getContentScope` to handle AnnotationPage items, as it should. I think that happened because the manifest was corrected after the PR was submitted. All unit tests pass now. 

Other points:

1. The purpose of TransformSet is to provide a set of translation, rotation, and scale transforms that conform to the IIIF spec expectation (where multiple transforms compound cumulatively, and the output state of each transform is the input state for the next transform), where (in a Three.js context) the three transforms could be provided to an Object3D to specify the local position/rotation/scale of the object correctly. So more than your vector example, it allows for:

```
object3D.position = tset.position
object3D.rotation = tset.rotation
object3D.scale = tset.scale
```

(I mainly use react-three-fiber currently so really it's `<GLTF position={tset.position} rotation={tset.rotation} scale={tset.scale}>` but still!)

2. Why does `combineTransformsToTRS` need to be rewritten to use quaternions? Happy to, just curious. 

3. Are you sure about `matrix.multiply(tmat)` needing to be changed to `premultiply`? My understanding was that since we want to apply each transform in order cumulatively, that this would apply the transforms in the order from the input array?

JulieWinchester committed:
Improve Geometry3d, fix issues, add new tests for transform matrix

JulieWinchester commented:
Okay @vincentmarchetti following our conversation during the IIIF Commons meeting today, I think I've now addressed all the remaining points:

* `combineTransformsToMatrix` is now correctly using `premultiply` for matrix multiplication. 
* `combineTransformsToTRS` multiples rotations using quaternions. 
* I've added a new test for combining multiple rotations, and added an entire new set of tests for `combineTransformstoMatrix`. As you suggested would happen, the behavior between the two methods is now identical in the tests covered, and these two approaches should now be equivalent. 

I think the PR is ready for further review now.

vincentmarchetti commented:
Have started reviewing. I am still  not sure if the operation on quaternions in combineTransformsToTRS should be multiply or premultiply; am working on graphics and test cases to convince all of us.

vincentmarchetti committed:
Added additional test case for a transform list that contains two RotateTransform instances
that do not commute

vincentmarchetti commented:
I think the function combineTransformsToTRS in Geometry3d.ts should be modified, the function applied to the quaternions near line 154 should be changed to premultiply, so that the final version would be:

```
const q1 = new Quaternion().setFromEuler(rotation);
const q2 = new Quaternion().setFromEuler(euler);
q1.premultiply(q2);
rotation.setFromQuaternion(q1, "XYZ");
```

While all of the unit tests succeed with the original code ( https://github.com/IIIF-Commons/manifesto-3d/pull/29/commits/23d2bf1b7e3d54525bfdd54aaf01c914b1e1c91f ), the test that combined two RotateTransforms, the tests with 'combines transform rotate -> rotate' , combine two RotateTransforms both about the x-axis, in this case the quaternion multiplication is independent of order (so multiply will give the same result are premultiply )

I prepared unit tests that combine a RotateTransform along the x axis with with along the y axis, in detail the transform tested is:

```
transform : [
    {
        type: "RotateTransform",
        y : 45
    },
    {
        type: "RotateTransform",
        x : 60
    }
]

```
To complete the unit test I am asserting that this combination of rotations is geometrically equal to :

```
transform : [
    {
        type: "RotateTransform",
        x : 60,
        y : 45
    }
]
```

I recognize that this is based on an interpretation of combining rotations inside a transform list, and that there may be a different consistent interpretation based on redefining the X and Y axis within the transform list in the same way that intrinsic axes are using within a single Euler rotation; but I can't see that explained in the three-js documentation.

However, making adding this unit test to the set for both the 'combineTransformsToMatrix' test suite and the 'combineTransformsToTRS'  under the description 'combines transform rotate -> rotate (2)' has results:
-- with current code , at commit   , the new unit test in the combineTransformsToMatrix suite, which is not sensitive to quaternion multiplication, succeeds;  while in the combineTransformsToTRS suite it fails.

-- when the quaternion calculation is changed to .premultiply, the new unit test succeeds in both suites; all other unit tests succeed.

I have not changed the code in Gerometry3d.ts, so at this commit on test should fail:
```
 1) iiif-3d tests
       Geometry3d: combineTransformsToTRS
         combines transform rotate -> rotate (2):
     AssertionError: expected 1.1831996401397158 to be close to 1.0471975511965976 +/- 0.000001
      at Context.<anonymous> (test/tests_3d/core_tests/Geometry3d.js:589:34)
      at process.processImmediate (node:internal/timers:478:21)

```

JulieWinchester committed:
Change quaternion multiply to premultiply

JulieWinchester commented:
@vincentmarchetti I agree with your assertion about the combination of rotation transforms and so does Three.js (used in the way the IIIF spec indicates separate transforms should be handled):

```
console.log('---TEST---')
v = new threejs_math.Vector3(0, 0, 1);
r1 = new threejs_math.Euler(0, 0.785398, 0);
r2 = new threejs_math.Euler(1.0472, 0, 0, 'XYZ');
v.applyEuler(r1);
v.applyEuler(r2);
console.log(v);

// Vector3 {
//   x: 0.7071066656470945,
//   y: -0.6123734015369311,
//   z: 0.35355194878198193
// }

v2 = new threejs_math.Vector3(0, 0, 1);
r3 = new threejs_math.Euler(1.0472, 0.785398, 0, 'XYZ');
v2.applyEuler(r3);
console.log(v2);

// Vector3 {
//   x: 0.7071066656470945,
//   y: -0.6123734015369309,
//   z: 0.3535519487819818
// }
```

I've changed the quaternion multiplication from `.multiply` to `.premultiply` as you suggested. All tests are passing now, including your added test. If you haven't seen any other issues and you see fit to merge, I can then try out pulling down latest manifesto and making any changes needed for that.
